ipq40xx: add initial support for Xiaomi Mesh Router D01

stas2z opened the pull request:
https://forum.openwrt.org/t/xiaomi-miwifi-mesh-router-d01-support/52820/21

adrianschmutzler commented:
Your Pull Request has formal issues, i.e. missing commit description and Signed-off-by. Note that there are special requirements to the commit description when adding a new device.

https://openwrt.org/submitting-patches

Despite, why is your support "initial"? I'd prefer "full" support.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -54,6 +54,10 @@ ipq40xx_setup_interfaces()
 		;;
+	xiaomi,meshd01)
+		ucidef_set_interface_lan "eth1 eth2 eth3"
+		ucidef_set_interface_wan "eth0"
```
ucidef_set_interfaces_lan_wan "eth1 eth2 eth3" "eth0"

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -84,6 +87,7 @@ case "$FIRMWARE" in
 	openmesh,a62 |\
 	qxwlan,e2600ac-c1 |\
 	qxwlan,e2600ac-c2 |\
+	xiaomi,meshd01 |\
```
Please take care of proper alphabetic sorting.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -155,6 +159,7 @@ case "$FIRMWARE" in
 	openmesh,a62 |\
 	qxwlan,e2600ac-c1 |\
 	qxwlan,e2600ac-c2 |\
+	xiaomi,meshd01 |\
```
Sorting.

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -9,6 +9,13 @@ preinit_set_mac_address() {
+		ip link set dev eth0 address "${base_mac}"
+		ip link set dev eth1 address $(macaddr_add "${base_mac}" 1)
+		ip link set dev eth2 address $(macaddr_add "${base_mac}" 2)
+		ip link set dev eth3 address $(macaddr_add "${base_mac}" 3)
```
Have you checked whether there are additional addresses in "0:ART", e.g. at 0x6 etc.?
If yes, please use them directly instead of doing arithmetics.

stas2z replied:
just checked, yes, it have, will fix it

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+
+/ {
+    model = "Xiaomi Mesh Router D01";
+    compatible = "xiaomi,meshd01", "qcom,ipq4019";
```
Based on your name, the compatible would be "xiaomi,mesh-router-d01".
In the forum, it looks to me like the correct name would be "xiaomi,miwifi-mesh-router-d01", though.

Please note that changing this will also require changes to all the cases above, DTS file name, device name in image/Makefile etc.

stas2z replied:
it'a miwifi mesh router (model d01 according to stock firmware) made by xiaomi
i can rename it, but i thought short xiaomi,meshd01 name will be enuff to identify this device

adrianschmutzler replied:
Please use the full name, this will be more consistent with the rest and also easier for the user to understand.

stas2z replied:
ive checked dts of other xiaomi boards
miwifi (all xiaomi routers are miwifi) is somwhere skipped and somewhere no
for example xiaomi,mir3g and xiaomi,miwifi-mini
"router" skipped always

so which full name i should use?

adrianschmutzler replied:
I'm not sure. BTW, I also wonder what to do with the d01. If it specifies a "submodel" of the mesh router, it should be included in the name. If it's just another naming scheme for the same device, then we shouldn't include it. So, I'd go for (at my current level of understanding)
xiaomi,miwifi-mesh
or
xiaomi,miwifi-mesh-d01

mir3g is older, collapsed names were more frequent back then. BTW, the "r" stands for router there, so technically it's included ;-)

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+    compatible = "xiaomi,meshd01", "qcom,ipq4019";
+
+    chosen {
+	bootargs-append = " root=/dev/ubiblock0_0 clk_ignore_unused";
```
Indent should be tabs only. This whole file seems mixed.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+    };
+
+    aliases {
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+	serial0 = &blsp1_uart1;
+    };
+
+
```
One empty line is enough.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+	    gpios = <&tlmm 40 GPIO_ACTIVE_LOW>;
+	};
+    };
+    soc {
```
Missing empty line before this node.

Please take care of similar formatting issues below, I won't mark them again ...

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+
+&nand {
+    compatible = "qcom,ipq4019-nand";
+    //reg = <0x7980000 0x1000>;
```
If it isn't needed, remove it.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+};
+
+
+
```
Also remove those empty lines at the end.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-meshd01.dts:
```diff
@@ -0,0 +1,390 @@
+&wifi1 {
+    pinctrl-0 = <&wifi1_pins>;
+    pinctrl-names = "default";
+    status = "okay";
```
status should always be first property in the node.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -198,6 +198,21 @@ define Device/cilab_meshpoint-one
 endef
 TARGET_DEVICES += cilab_meshpoint-one
 
+define Device/xiaomi_meshd01
```
As discussed in the DTS, you will have to adjust this to the new compatible, just with "_" instead of ",".

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -198,6 +198,21 @@ define Device/cilab_meshpoint-one
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Xiaomi
+	DEVICE_MODEL := Mesh D01
```
This should correspond to the model string in DTS.

Despite, what exactly is the D01? If it is some kind of hardware revision, it might be more appropriate to use
`DEVICE_VARIANT := D01`

stas2z replied:
it's not a revision i think, more like internal model name
miwifi_d01_firmware_52359_1.0.68.bin - is a stock fw file name
same with other xiaomi routers, for example router3 pro: miwifi_r3p_firmware_f5d80_2.16.29.bin

adrianschmutzler replied:
I see. So, not DEVICE_VARIANT needed.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -198,6 +198,21 @@ define Device/cilab_meshpoint-one
+	DEVICE_VENDOR := Xiaomi
+	DEVICE_MODEL := Mesh D01
+	SOC := qcom-ipq4019
+	DEVICE_DTS := qcom-ipq4019-meshd01
```
That's not required if you set SOC and name your DTS file correctly.

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -198,6 +198,21 @@ define Device/cilab_meshpoint-one
+endef
+TARGET_DEVICES += xiaomi_meshd01
+
+
```
One empty line only.

stas2z commented:
> Your Pull Request has formal issues, i.e. missing commit description and Signed-off-by. Note that there are special requirements to the commit description when adding a new device.
> 
> https://openwrt.org/submitting-patches
> 
> Despite, why is your support "initial"? I'd prefer "full" support.



> Your Pull Request has formal issues, i.e. missing commit description and Signed-off-by. Note that there are special requirements to the commit description when adding a new device.
> 
> https://openwrt.org/submitting-patches
> 
> Despite, why is your support "initial"? I'd prefer "full" support.
ok, it's my first commit so i can miss some formal details
also im not an experienced git user, so i don't know is it possible to add signoff to existing commits or i need to reset, fix all issues and commit it with fixes and signoff?

about full support, i see several problems with it, at least i don't know what to do with built-in plc and spi driven leds here, thats why it's initial. But generally all common hardware seems like work properly.

adrianschmutzler commented:
> about full support, i see several problems with it, at least i don't know what to do with built-in plc and spi driven leds here, thats why it's initial. But generally all common hardware seems like work properly.

Then please provide details about that in the commit message. You explain the shortcomings of your submissing and add something like "Known issues" or "Unsupported features" there.

stas2z commented:
Sorry, git is not my friend
Ive tried to fix all issues here, but i don't know, probably it's better to create new pull request with one commit and signoff?

adrianschmutzler commented:
Please no new PR, we loose all the history every time. Just try to change/squash your existing commits with "git rebase -i ..." and then force-push "git push -f ..." to your GitHub repo again.

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -22,6 +22,12 @@ preinit_set_mac_address() {
+	xiaomi,meshd01)
+		ip link set dev eth0 address $(mtd_get_mac_binary "0:ART" 0x0)
+		ip link set dev eth1 address $(mtd_get_mac_binary "0:ART" 0x6)
+		ip link set dev eth2 address $(mtd_get_mac_binary "0:ART" 0xC)
```
I'd prefer lower-case: 0xc.
Have you checked that the assignment corresponds to stock firmware?

stas2z replied:
yes, stock firmware extract this same way using dump of ART and hexdump :)

`
get_eth0()
{
	eth0adr="`hexdump -s 0 -n 6 -e '6/1 "%02x:"'  /dev/"$art_mtd" | cut -b-17`"
}

get_eth1()
{
	eth1adr="`hexdump -s 6 -n 6 -e '6/1 "%02x:"'  /dev/"$art_mtd" | cut -b-17`"
}

get_eth2()
{
	eth2adr="`hexdump -s 12 -n 6 -e '6/1 "%02x:"'  /dev/"$art_mtd" | cut -b-17`"
}

get_eth3()
{
	eth3adr="`hexdump -s 18 -n 6 -e '6/1 "%02x:"'  /dev/"$art_mtd" | cut -b-17`"
}
`

stas2z commented:
Ok, seems like i was able to force push

adrianschmutzler commented:
> Ok, seems like i was able to force push

Yes, this looks good. When you push with the next set of changes, remove the full stop at the end of the title.

stas2z commented:
Oh, full stop is a finishing dot, yes?

adrianschmutzler commented:
Yes, full stop, dot, point, period, all the same :-)

stas2z commented:
Ok, ive changed device name do miwifi-mesh-d01 (probably xiaomi will produce more mesh oriented devices in future) and tested renamed image with my device, seems like everything done properly

stas2z committed:
ipq40xx: add support for Xiaomi Mesh D01 router

Xiaomi Mesh D01 router is an IPQ4019 device.
It equipped with some special features like built-in PLC and bluetooth which are not supported by mainline kernel yet.
All common devices are working (CPU/Ethernet/WIFI/NAND/Serial) except spi based RGB led controller.

eth0/eth1/eth2 are ethernet ports
eth3 assigned to built-in PLC

Upgrading from stock firmware is not trivial yet. No developer firmware provided by Xiaomi, so no way to get ssh or telnet oficially. Device contains serial ttl header with pinout silkscreened.

Signed-off-by: Alexey Golyshin <stas2z@gmail.com>

stas2z committed:
ipq40xx: mifiwi-mesh-d01 support update

Add support for uboot-envtools
Remove USB init from dts (device have no USB ports)

Signed-off-by: Alexey Golyshin <stas2z@gmail.com>

stas2z committed:
Merge branch 'master' into master

adrianschmutzler commented:
Your recent commits look strange. Anyway, please squash your (valid) changes into a single proper commit again.

adrianschmutzler commented:
This needs to move DTS files to "files" instead of "files-4.19" and also needs to be updated for kernel 5.4. Since the last commit is from February and there has been no response to my request from one month ago, I'm closing this now.
Please request reopening when you plan to continue (_before_ pushing anything to your GitHub branch).

Neustradamus commented:
@stas2z: Any news about your contribution?

stas2z commented:
> @stas2z: Any news about your contribution?

Ive tried but was ignored for a long time, i have no reason and time to return, update tree and trying to commit it again

stas2z commented:
if anyone still interested, here is the forked openwrt tree with all changes i made to support d01 based on current (atm) trunk openwrt tree
https://github.com/stas2z/openwrt
ive built it and tested on my device

Neustradamus commented:
@stas2z: Nice, maybe you can create a new PR?
